other/perun-propagate/etc/perun_propagate

```diff
@@ -43,37 +56,111 @@ perun_propagate_start() {
 	else
 		# MIT Kerberos
-		PRINCIPAL=`echo -e "read_kt ${KEYTAB}\nlist" | ktutil list 2>/dev/null | awk '($3~"^host/") && ($3!~"\\\.ipv6\\\.") { print $3 }' | head -1`
-		kinit -k -t "${KEYTAB}" "${PRINCIPAL}"
+		PRINCIPALS=`echo -e "read_kt ${KEYTAB}\nlist" | ktutil list 2>/dev/null | awk '($3~"^host/") && ($3!~"\\\.ipv6\\\.") { print $3 }' 2>/dev/null | sort | uniq `
+		KERBEROS_TYPE="mit"
 	fi
 
+	#For every unique Principal do kinit and start perun-propagate
 	set +e
-	if which timeout stdbuf &>/dev/null; then
-		OUT=`timeout ${TIMEOUT} stdbuf -oL -eL remctl "${PERUN_SERVER}" perun propagate 2>&1`
-		EXIT_CODE=$?
-	else
-		OUT=`remctl "${PERUN_SERVER}" perun propagate 2>&1`
-		EXIT_CODE=$?
+	ITERATOR=0;
+	while read -r LINE; do
+		if [ "${KERBEROS_TYPE}" = "heimdal" ]; then
+			kinit -k -t "${KEYTAB}" "${LINE}"
+		else	
+			kinit --no-afslog --keytab="${KEYTAB}" "${LINE}"
+		fi
+		TMP_ERROR=`mktemp perun-propagate.XXXXXXXXX`
+		remctl "${PERUN_SERVER}" perun propagate 2>&1 | tee ${TMP_ERROR} >> ${LOG} &
+		PID=$!
+		PID_ARRAY[$ITERATOR]=$PID
+		TMP_ERROR_ARRAY[$ITERATOR]=$TMP_ERROR
+		PRINCIPAL_ARRAY[$ITERATOR]=$LINE
+	done <<< "$PRINCIPALS"
+
+	#Start time of propagation in seconds
+	START_TIME=`date +%s`
+	#Prepare firt run of while with same START_TIME
+	NOW_TIME=$START_TIME
+	#Count of all processes in PIDS_ARRAY to know how much of them still running
+	PROCESSES_COUNT=${#PID_ARRAY[@]}
+	#Waiting for all running processes but not more than TIMEOUT in seconds
+	while [ `expr $NOW_TIME - $START_TIME` -lt "$TIMEOUT" ]; do
+		DONE_PROCESSES=0
+		for i in ${!PID_ARRAY[*]}; do
+			if is_process_running ${PID_ARRAY[$i]} ; then
+				break
+			else
+				DONE_PROCESSES=`expr $DONE_PROCESSES + 1`
+			fi
+		done
+		#All processes ended so we can continue
+		if [ "$DONE_PROCESSES" -eq "$PROCESSES_COUNT" ]; then
+			break
+		fi
+		NOW_TIME=`date +%s`
+	done
+	#Get all statuses and errors and prepare messages and exit code
+	for i in ${!PID_ARRAY[*]}; do
+		is_process_running ${PID_ARRAY[$i]}
+		#process $i is still running
+		if [ "$?" = 0 ]; then
+			#if still running, we want to belive this is ok so we set exit code to 0 if not in error yet
+			OUT="Perun_propagate for host ${PRINCIPAL_ARRAY[$i]} still running even after timeout.\n"
+			if [ "${EXIT_CODE}" != 1 ]; then
+				EXIT_CODE=0
+			fi
+		else
+			#script ends so we need to get info about exit code of this script
+			WAITING_ID=`expr $i + 1`
+			wait %$WAITING_ID
+			SCRIPT_EXIT_CODE=$?
+			if [ "$SCRIPT_EXIT_CODE" = 0 ]; then
+				#scripts ends without error, so just add info from TMP FILE and set exit code to 0 if not in error yet
+				OUT="Perun_propagate for host ${PRINCIPAL_ARRAY[$i]} ends with status 0.\n"
+				if [ "${EXIT_CODE}" != 1 ]; then
+					EXIT_CODE=0
+				fi
+			else
+				#script ends with error so we need to get info if this error is just a problem with host or some other problem
+				grep -q "doesn't have assigned any facility" "${TMP_ERROR_ARRAY[$i]}"
```

Comment:
If it cannot be done differently, I propose to set the text "doesn't have assigned any facility" as a constant at the beginning of the script. In case of change of the error message it will be little bit easier to find out why the condition isn't working.